File: R/994-rotting-oranges/main.go
package main

import "fmt"

func orangesRotting(grid [][]int) int {
	if len(grid) == 0 || len(grid[0]) == 0 {
		return 0
	}
	filter := func(grid [][]int) ([][]int, bool) {
		var pos [][]int
		fresh := false
		for i := 0; i < len(grid); i++ {
			for j := 0; j < len(grid[0]); j++ {
				if grid[i][j] == 2 {
					pos = append(pos, []int{i, j})
				}
				if grid[i][j] == 1 {
					fresh = true
				}
			}
		}
		return pos, fresh
	}
	mark := func(grid [][]int, candi [][]int) {
		for i := 0; i < len(candi); i++ {
			pos := candi[i]
			grid[pos[0]][pos[1]] = 0
			if pos[0] > 0 && grid[pos[0]-1][pos[1]] == 1 {
				grid[pos[0]-1][pos[1]] = 2
			}
			if pos[0] < len(grid)-1 && grid[pos[0]+1][pos[1]] == 1 {
				grid[pos[0]+1][pos[1]] = 2
			}
			if pos[1] > 0 && grid[pos[0]][pos[1]-1] == 1 {
				grid[pos[0]][pos[1]-1] = 2
			}
			if pos[1] < len(grid[0])-1 && grid[pos[0]][pos[1]+1] == 1 {
				grid[pos[0]][pos[1]+1] = 2
			}
		}
	}

	candi, fresh := filter(grid)
	if len(candi) == 0 && !fresh {
		return 0
	}
	if len(candi) == 0 && fresh {
		return -1
	}
	if len(candi) > 0 && !fresh {
		return 0
	}

	times := 0
	for {
		mark(grid, candi)
		times++
		candi, fresh = filter(grid)
		if len(candi) == 0 && !fresh {
			return times - 1
		}
		if len(candi) == 0 && fresh {
			return -1
		}
		if len(candi) > 0 && !fresh {
			return times
		}
	}
}

func main() {
	fmt.Println(orangesRotting([][]int{{2, 1, 1}, {1, 1, 0}, {0, 1, 1}}))
}

File: R/994-rotting-oranges/main.rs
struct Solution {}

use std::collections::VecDeque;
impl Solution {
  pub fn oranges_rotting(grid: Vec<Vec<i32>>) -> i32 {
    let mut grid = grid;
    let mut q: VecDeque<(usize, usize, i32)> = VecDeque::new();

    let mut fresh: bool = false;
    (0..grid.len()).for_each(|row| {
      (0..grid[0].len()).for_each(|col| {
        if grid[row][col] == 2 {
          q.push_back((row, col, 0));
        }
        if grid[row][col] == 1 {
          fresh = true;
        }
      });
    });

    if !fresh {
      return 0;
    }

    if q.len() == 0 {
      return -1;
    }

    let mut ans: i32 = 0;

    while q.len() > 0 {
      let (row, col, step) = q.pop_front().unwrap();
      if step > ans {
        ans = step;
      }
      if row > 0 && grid[row - 1][col] == 1 {
        grid[row - 1][col] = 2;

        q.push_back((row - 1, col, step + 1));
      }
      if row < grid.len() - 1 && grid[row + 1][col] == 1 {
        grid[row + 1][col] = 2;
        q.push_back((row + 1, col, step + 1));
      }
      if col > 0 && grid[row][col - 1] == 1 {
        grid[row][col - 1] = 2;
        q.push_back((row, col - 1, step + 1));
      }
      if col < grid[0].len() - 1 && grid[row][col + 1] == 1 {
        grid[row][col + 1] = 2;
        q.push_back((row, col + 1, step + 1));
      }
    }

    fresh = false;
    (0..grid.len()).for_each(|row| {
      (0..grid[0].len()).for_each(|col| {
        if grid[row][col] == 1 {
          fresh = true;
        }
      });
    });

    match fresh {
      true => -1,
      _ => ans,
    }
  }
}
